nav2_behaviors/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2022 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+
+#include "assisted_teleop.hpp"
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_behaviors
+{
+AssistedTeleop::AssistedTeleop()
+: TimedBehavior<AssistedTeleopAction>(),
+  feedback_(std::make_shared<AssistedTeleopAction::Feedback>())
+{}
+
+void AssistedTeleop::onConfigure()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // set up parameters
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "projection_time", rclcpp::ParameterValue(1.0));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "simulation_time_step", rclcpp::ParameterValue(0.1));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "teleop_cmd_vel", rclcpp::ParameterValue(std::string("teleop_cmd_vel")));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "joystick_topic", rclcpp::ParameterValue(std::string("joystick_topic")));
+
+  node->get_parameter("projection_time", projection_time_);
+  node->get_parameter("simulation_time_step", simulation_time_step_);
+
+  std::string teleop_cmd_vel;
+  node->get_parameter("teleop_cmd_vel", teleop_cmd_vel);
+  node->get_parameter("joystick_topic", joystick_topic_);
+
+  vel_sub_ = node->create_subscription<geometry_msgs::msg::Twist>(
+    teleop_cmd_vel, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::teleopVelocityCallback,
+      this, std::placeholders::_1));
+
+  joy_sub_ = node->create_subscription<sensor_msgs::msg::Joy>(
```

Comment:
We may need to have another sub for thing, but that cannot be a joystick, because again we have no standard mapping for what buttons mean what for any particular user.